packages/hardhat-core/src/internal/solidity/compiler/downloader.ts

```diff
@@ -159,7 +159,19 @@ export class CompilerDownloader {
       await this.downloadCompilersList(platform);
     }
 
-    return fsExtra.readJson(this.getCompilersListPath(platform));
+    try {
+      return await fsExtra.readJSON(this.getCompilersListPath(platform));
+    } catch (error) {
+      // if parsing throws a syntax error, redownload and parse once more
+      if (!(error instanceof SyntaxError)) {
+        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
+        throw error;
+      }
+
+      await this.downloadCompilersList(platform);
+
+      return await fsExtra.readJSON(this.getCompilersListPath(platform));
```

Comment:
One alternative here is to do just `return getCompilersList(platform)`. That feels better to me, but if we do that we should also include a `retryLimit` variable.

Reply:
Something like this:

```typescript
  public async getCompilersList(
    platform: CompilerPlatform,
    pendingRetries = 3
  ): Promise<CompilersList> {
    if (!(await this.compilersListExists(platform))) {
      await this.downloadCompilersList(platform);
    }

    try {
      return await fsExtra.readJSON(this.getCompilersListPath(platform));
    } catch (error) {
      // if parsing throws a syntax error, redownload and parse once more
      if (!(error instanceof SyntaxError) || pendingRetries === 0) {
        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
        throw error;
      }

      // remove the malformed list and retry
      await fsExtra.remove(this.getCompilersListPath(platform));
      return this.getCompilersList(platform, pendingRetries - 1);
    }
  }
```